code/go/0chain.net/blobbercore/handler/grpc_handler_integration_test.go

```diff
@@ -0,0 +1,283 @@
+package handler
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/spf13/viper"
+
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
+
+	"0chain.net/blobbercore/config"
+	"google.golang.org/grpc"
+
+	"0chain.net/blobbercore/blobbergrpc"
+)
+
+const BlobberAddr = "localhost:7031"
+const RetryAttempts = 8
+const RetryTimeout = 3
+
+func TestBlobberGRPCService_IntegrationTest(t *testing.T) {
+	args := make(map[string]bool)
+	for _, arg := range os.Args {
+		args[arg] = true
+	}
+	if !args["integration"] {
+		t.Skip()
+	}
+
+	ctx := context.Background()
+
+	var conn *grpc.ClientConn
+	var err error
+	for i := 0; i < RetryAttempts; i++ {
+		log.Println("Connection attempt - " + fmt.Sprint(i+1))
+		conn, err = grpc.Dial(BlobberAddr, grpc.WithInsecure())
+		if err != nil {
+			log.Println(err)
+			<-time.After(time.Second * RetryTimeout)
+			continue
+		}
+		break
+	}
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer conn.Close()
+	blobberClient := blobbergrpc.NewBlobberClient(conn)
+
+	pwd, err := os.Getwd()
+	if err != nil {
+		t.Fatal(err)
+	}
+	configDir := strings.Split(pwd, "/code/go")[0] + "/config"
+	config.SetupDefaultConfig()
+	config.SetupConfig(configDir)
+	config.Configuration.DBHost = "localhost"
+	config.Configuration.DBName = viper.GetString("db.name")
+	config.Configuration.DBPort = viper.GetString("db.port")
+	config.Configuration.DBUserName = viper.GetString("db.user")
+	config.Configuration.DBPassword = viper.GetString("db.password")
+	db, err := gorm.Open(postgres.Open(fmt.Sprintf(
+		"host=%v port=%v user=%v dbname=%v password=%v sslmode=disable",
+		config.Configuration.DBHost, config.Configuration.DBPort,
+		config.Configuration.DBUserName, config.Configuration.DBName,
+		config.Configuration.DBPassword)), &gorm.Config{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	tdController := NewTestDataController(db)
+
+	t.Run("TestGetAllocation", func(t *testing.T) {
+		err := tdController.ClearDatabase()
```

Comment:
Most new 0chain unit tests by @MurashovVen are written as [table-driven tests](https://blog.golang.org/subtests#TOC_2.). I think it would be a good idea to adopt that style here.

Also, can we have all tests running in parallel. 
```go
t.Run(...., ... {
    t.Parallel()
```

This is to help detect [race errors](https://golang.org/doc/articles/race_detector). To avoid confounding race errors run the unit test with the race flag, `go test -race` and remove race errors caused by the test and raise issues for race errors in the production code. Hint: try to copy by value to avoid unnecessarily sharing memory addresses between tests.

Reply:
This is not a unit test - its an integration test. Its testing the api.

Reply:
I have reformatted it to be table driven tests.

Reply:
Handling race errors would be alot easier through unit tests, this is however integration tests.

If we were testing crud api, imagine the delete happening before the object was created. To solve this we will be forced to create large number of rows in tables such that its deleting something that already existed in db and wasnt created by api, it will be hard to tell which row is being used by which tests.

Since unit test does not even deal with db and doesnt care about db state it will be much easier to handle it there.